modules/gcp/storage.go

```diff
@@ -0,0 +1,214 @@
+	bucket := client.Bucket(name)
+
+	// Creates the new bucket.
+	return bucket.Create(ctx, projectID, attr)
+}
+
+// DeleteStorageBucket destroys the Google Storage bucket.
+func DeleteStorageBucket(t *testing.T, name string) {
+	err := DeleteStorageBucketE(t, name)
+	if err != nil {
+		t.Fatal(err)
+	}
+}
+
+// DeleteStorageBucketE destroys the S3 bucket in the given region with the given name.
+func DeleteStorageBucketE(t *testing.T, name string) error {
+	logger.Logf(t, "Deleting bucket %s", name)
+
+	ctx := context.Background()
+
+	client, err := storage.NewClient(ctx)
+	if err != nil {
+		return err
+	}
+
+	return client.Bucket(name).Delete(ctx)
+}
+
+// ReadBucketObject reads an object from the given Storage Bucket and returns its contents.
+func ReadBucketObject(t *testing.T, bucketName string, filePath string) io.Reader {
+	out, err := ReadBucketObjectE(t, bucketName, filePath)
+	if err != nil {
+		t.Fatal(err)
+	}
+	return out
+}
+
+// ReadBucketObjectE reads an object from the given Storage Bucket and returns its contents.
+func ReadBucketObjectE(t *testing.T, bucketName string, filePath string) (io.Reader, error) {
+	logger.Logf(t, "Reading object from bucket %s using path %s", bucketName, filePath)
+
+	ctx := context.Background()
+
+	client, err := storage.NewClient(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	bucket := client.Bucket(bucketName)
+	r, err := bucket.Object(filePath).NewReader(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	return r, nil
+}
+
+// WriteBucketObject writes an object to the given Storage Bucket and returns its URL.
+func WriteBucketObject(t *testing.T, bucketName string, filePath string, body io.Reader, contentType string) string {
+	out, err := WriteBucketObjectE(t, bucketName, filePath, body, contentType)
+	if err != nil {
+		t.Fatal(err)
+	}
+	return out
+}
+
+// WriteBucketObjectE writes an object to the given Storage Bucket and returns its URL.
+func WriteBucketObjectE(t *testing.T, bucketName string, filePath string, body io.Reader, contentType string) (string, error) {
+	// set a default content type
+	if contentType == "" {
+		contentType = "application/octet-stream"
+	}
+
+	logger.Logf(t, "Writing object to bucket %s using path %s and content type %s", bucketName, filePath, contentType)
+
+	ctx := context.Background()
+
+	client, err := storage.NewClient(ctx)
+	if err != nil {
+		return "", err
+	}
+
+	w := client.Bucket(bucketName).Object(filePath).NewWriter(ctx)
+	w.ContentType = contentType
+
+	// Don't set any ACL or cache control properties for now
+	//w.ACL = []storage.ACLRule{{Entity: storage.AllAuthenticatedUsers, Role: storage.RoleReader}}
+	// set a default cache control (1 day)
+	//w.CacheControl = "public, max-age=86400"
+
+	if _, err := io.Copy(w, body); err != nil {
+		return "", err
+	}
+	if err := w.Close(); err != nil {
+		return "", err
+	}
+
+	const publicURL = "https://storage.googleapis.com/%s/%s"
+	return fmt.Sprintf(publicURL, bucketName, filePath), nil
```

Comment:
Does it make more sense to return the bucket object itself and have a separate `GetBucketUrl(bucketObject)` function?

Reply:
I would argue in that case it should return only the error object as you pass in the `io.Reader` as a parameter. Let me know your thoughts and I'll adjust accordingly.

Reply:
Oh, interesting. We can just leave as is.

Reply:
ok